Update contribution guides

DenisGorbachev opened the pull request:
* [X] Put an X in this bracket to confirm you have read the
  [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).

## Summary

> This PR follows the discussion in "[Merge contribution guides?](https://leanprover.zulipchat.com/#narrow/stream/270676-lean4/topic/Merge.20contribution.20guides.3F)".
> This PR's submission was approved by @semorrison.

Currently, there are two guides: [CONTRIBUTING.md](https://github.com/leanprover/lean4/blob/master/CONTRIBUTING.md) and [doc/contributions.md](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).

It would be better to merge the guides into a single CONTRIBUTING.md, and link to that file from PULL_REQUEST_TEMPLATE.md.

## Notes

I tried to preserve the original wording, but I've shortened some paragraphs & removed ambiguous sentences. Please correct if wrong.

For review, it may be easier to read the final files instead of the diffs.

## Questions

* What should we write in "Feature requests"? (I added my proposal directly to the file)
* Should we keep "Asking questions" section? 
  * In the "Asking questions" section: should we keep the following sentence: "A lot of people report feeling that they are "wasting expert time", but nobody on the team feels this way"? (I propose to keep since it encourages people to reach out)
* In "Making commits" section, there is a sentence: "Check that your changes do not introduce performance regressions". How exactly should the contributors check for performance regressions?

semorrison commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,12 @@
 
-* You can manage the `awaiting-review`, `awaiting-author`, and `WIP` labels
-  yourself, by writing a comment containing one of these labels on its own line.
+Parent issue:
```
This is confusing. What is the "parent" issue? Presumably it is about the "Add the link to your `RFC` or `bug` issue below"? But why separate them out?

semorrison replied:
I see. How about we change this section title to "Read and remove this section before submitting"?

semorrison commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,12 @@
 
-* If that issue does not already have approval from a developer,
-  please be sure to open this PR in "Draft" mode.
+P.S. You can manage the `awaiting-review`, `awaiting-author`, and `WIP` labels yourself, by writing a comment containing one of these labels on its own line.
```
Don't like "P.S.", this isn't a letter!

DenisGorbachev replied:
Removed

semorrison commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,12 @@
-  PRs missing this link will be marked as `missing RFC`.
+* Ensure your PR follows the [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/CONTRIBUTING.md).
+* Add the link to your `RFC` or `bug` issue below.
+* If the issue does not already have approval from a developer: submit the PR as draft.
```
colon is weirdly ungrammatical?

DenisGorbachev replied:
Changed to comma, sorry

semorrison commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,12 @@
-  please be sure to open this PR in "Draft" mode.
+P.S. You can manage the `awaiting-review`, `awaiting-author`, and `WIP` labels yourself, by writing a comment containing one of these labels on its own line.
 
-* Please make sure the PR has excellent documentation and tests.
-  If we label it `missing documentation` or `missing tests` then it needs fixing!
```
Where did this go?

DenisGorbachev replied:
* "Please make sure the PR has excellent documentation and tests" - it's in the "External Contribution Guidelines", mentioned on the first line. Should we emphasize it again here?
* "If we label it `missing documentation` or `missing tests` then it needs fixing!" - seems obvious, should we state it explicitly?

semorrison commented on .github/PULL_REQUEST_TEMPLATE.md:
```diff
@@ -1,14 +1,12 @@
-* [ ] Put an X in this bracket to confirm you have read the
-  [External Contribution Guidelines](https://github.com/leanprover/lean4/blob/master/doc/contributions.md).
```
Why did you remove the checkbox?

DenisGorbachev replied:
The whole "Before submitting" section is supposed to be removed, so there's no need to check the checkbox.

I thought it would be better to ask the submitters to remove the "Before submitting" section because of two reasons:

* Less text to read for maintainers
* Less repetitive actions for submitters (e.g. for people who are already familiar with the contribution guidelines)

semorrison commented:
Too many arbitrary changes that would need discussion.

Perhaps you could try again with the absolute minimal changes required to merge the files? Otherwise we will get to it: even the report about the duplication is very helpful.

DenisGorbachev commented:
@semorrison I'd love to make minimal changes, but the sentences are interrelated - can't change one without changing the other.

I've fixed some points that you've mentioned and replied to others.

Sorry if I did something wrong - please let me know if it needs more improvements.

semorrison commented:
The basic problem is that doc/contributions.md is the new version that should take priority, and was recently written by Leo. By merging it into the other document is has essentially been lost.

I think the only viable path to combining these documents is:
* `doc/contributions.md` stays exactly as is, except moved to `CONTRIBUTING.md`
* anything currently in `CONTRIBUTING.md` which is genuinely additional to anything mentioned in `docs/contributions.md` could perhaps just be added at the end.

But really anything that is much beyond deleting `CONTRIBUTING.md` and replacing it with `docs/contributions.md` unmodifed is going to require too much editorial input to be viable via PR.

DenisGorbachev commented:
Updated https://github.com/DenisGorbachev/lean4/blob/contribution-guidelines/CONTRIBUTING.md:

* Fully reused doc/contributions.md
* Added "Helpful links" section
* Replaced:
  * Old: `In the past, we accepted most pull requests. This practice produced hard to maintain code, performance problems, and bugs. In order to improve the quality and maintainability of our codebase, we've established the following guidelines for external contributions.`
  * New: `Thank you for your interest in contributing to Lean! There are many ways to contribute and we appreciate all of them.`

Reason for replacement: since CONTRIBUTING.md should help people who want to contribute, it's better to start with a welcome message.

---

Is it correct?

If yes - should I open a new PR or would you like to reopen the current one?

semorrison commented:
Please revert the change from the "replaced:" bullet point. That's a significant editorial change!

DenisGorbachev commented:
Done

DenisGorbachev commented:
awaiting-review

(commenting to change the label)

DenisGorbachev commented:
@semorrison Is there anything else that should be done on this PR?

semorrison commented:
In principle I'm happy now, but the problem is that there is a slide deck for the community meeting that uses the old URL, so I'm very hesitant to merge now. :-)

semorrison commented:
Sorry, on further reflection the right solution is to leave `docs/contributions.md` in place (it becomes part of the online manual, there are links to it, etc. etc.), and instead to just delete the contents of `CONTRIBUTING.md` and replace it with "Please see `docs/contribuctions.md`."

DenisGorbachev commented:
Actually, https://lean-lang.org/lean4/doc/contributions.html returns 404. This is because `doc/contributions.md` is never mentioned in `doc/SUMMARY.md`, so it is never rendered by mdBook. Also, in the whole `lean4` repo, there are two references to `doc/contributions.md`: in `README.md` and in `PULL_REQUEST_TEMPLATE.md` (both references are fixed in the PR).

Meanwhile, `CONTRIBUTING.md` is linked to by GitHub on every new PR page (that's just how GitHub works).

Let me know if you'd still like to move the text back to `doc/contributions.md` or keep it in `CONTRIBUTING.md`.

DenisGorbachev commented:
Also, is it OK to provide additional info (like I've done in the previous comment), or would it be better to just implement your suggestions immediately?

semorrison commented:
Thanks for your patience on this one, @DenisGorbachev! Just checking what we want to do here.

DenisGorbachev commented:
@semorrison Is there any way I can help regarding this PR?

DenisGorbachev commented:
@semorrison Would you like me to do anything else on this PR?
